contrib/ros/envs/ros_env.py

```diff
@@ -1,41 +1,164 @@
+from geometry_msgs.msg import Pose, Point
 import numpy as np
-import rospy
 
 from rllab.core.serializable import Serializable
-from rllab.envs.base import Env
+from rllab.envs.base import Env, Step
 from rllab.misc.ext import get_seed
+from rllab.spaces import Box
+
+from contrib.ros.util.common import rate_limited
+from contrib.ros.util.gazebo import Gazebo
 
 
 class RosEnv(Env, Serializable):
     """
     Superclass for all ros environment
     """
 
-    def __init__(self):
+    def __init__(self, task_obj_mgr, robot, simulated=False):
```

Comment:
If you are going to pass a `robot` to `RosEnv`, who is responsible for its lifecycle?  Does the `robot` need to be initialized before passed to RosEnv? Can I call functions on `robot` after passing it to `RosEnv`, or will that corrupt `RosEnv`'s view of the world state?

Why not do what you did before: a base environment for each robot and subclasses of that environment for specific tasks? Perhaps you would like to try the same task with different robots?

Reply:
I think we don't need a base environment for each robot. In RosEnv, it will use the functions of generative robot interface, so it will still be a generative base env. And I would like to give every specific robot's specific task a individual sub env class so that they can have different sample_goal, get_observation, reward, etc. functions. And the robot need to be initialized in that sub env for a specific robot's specific task.

Reply:
Which means the public interfaces that the base robot class asks every sub class to implement. Here we ask every specific robot interface to implement reset, send_command, action_space, get_observation, observation_space, and they are what the RosEnv will use. So RosEnv actually doesn't know what's the actual robot he is dealing with.

Reply:
I think the best way to convince is to add more different robots...

Reply:
I think the crucial question here is whether the RosEnv is too specific now. But I think it's not, as he is not assuming any specific robot or task.

Reply:
I'm closing this pull request and will open a new one for new change.